### PR TITLE
Enhancement: Add Sub-Resourcing

### PR DESCRIPTION
**Pull request by mortedecai**
Adds the ability to have sub-resources on the root resource or instantiated resources.

This PR also takes care of properly routing incoming requests to the longest match, as opposed to the order added.

The `hello` integration test has been updated to have an instanced sub-resource.  The `todo` integration test has not been updated at this point as these changes will be extensive to provide a, non-perfect, example of using both instanced and non-instanced sub resources.

### CODE REVIEW

File: api.go [hunk 1]
```diff
@@ -78,30 +83,58 @@ func (i ID) Find(s string) (string, bool) {
...
 	SetID(id ID) error
+	// SetHandler sets the handler function for this resource.
 	SetHandler(handler HandlerFunction)
+	// GetIDValue retrieves the ID value from the provided context for this call.
 	GetIDValue(ctx context.Context) (string, error)
+	// AddSubResource adds a sub-resource which does not depend on any particular instantiated resource instance.
+	//   For example, /users/search would allow the creation of searches across all users, but does not depend on any particular resource instantiations.
+	//
+	// Should the sub-resource need to access a particular resource instance, it MUST be added via AddInstancedSubResource.
+	AddSubResource(Resource) error
```
Comment (keithpaterson) [inline api.go]: could we call these Child resources instead?  I feel like that means the same thing but would be a bit more intuitive to the seasoned developer.

I think the Parent-Child relationship is intuitive and more common nomenclature

File: api.go [hunk 2]
```diff
@@ -78,30 +83,58 @@ func (i ID) Find(s string) (string, bool) {
...
+	// AddInstancedSubResource adds a sub-resource which depends on a particular instantiated resource instance.
+	//   For example:
+	//     * `/users/<id>/profile` could allow the viewing or editing of a particular users profile.
+	//     * `/users/<id>/emails` could allow the creation, viewing or editing of a particular users email(s).
+	//     * `/users/<id>/emails/foo%40bar.com` could allow the deletion of a particular users `foo@bar.com` e-mail.
+	AddInstancedSubResource(Resource) error
```
Comment (keithpaterson) [inline api.go]: yep I'm picking on naming.

How about DependentChildResource here (and maybe IndependentChildResource above)?  Or just Child for this (which I suspect will be the more common case) and IndependentChild for the previous function?

.. because Child and BastardChild would be fun and yet somehow inappropriate.

File: api.go [hunk 3]
```diff
@@ -78,30 +83,58 @@ func (i ID) Find(s string) (string, bool) {
...
 	LogHolder
-	name      ResourceName
-	actionMap actionFuncMap
-	id        ID
-	handler   HandlerFunction
+	name                  ResourceName
+	actionMap             actionFuncMap
+	id                    ID
+	handler               HandlerFunction
+	subResources          ResourceMap
```
Comment (keithpaterson) [inline api.go]: obviously whatever we agree on re: naming would be reflected here as well

File: api.go [hunk 4]
```diff
@@ -228,12 +287,40 @@ func (bar *BaseAPIRes) SetHandler(handler HandlerFunction) {
 	bar.handler = handler
 }
 
+func (bar *BaseAPIRes) findSubResource(c context.Context, w http.ResponseWriter, req *http.Request) (Resource, error) {
+	segments := c.Value(KeyURISegments).([]ResourceName)
+	if len(segments) == 0 {
+		// TODO: should this just return bar?
```
Comment (keithpaterson) [inline api.go]: Answer: No, bar is not a sub-resource of bar so it wouldn't make sense to return it here.

File: host.go [hunk 5]
```diff
@@ -80,16 +81,34 @@ func (h *host) GetResource(name ResourceName) (res Resource, found bool) {
...
 	reqPaths := ResourceNames(pathSegs)
 	ctx := req.Context()
+	pathIdx := 0
+	leadSlash := false
+	if strings.HasPrefix(req.URL.Path, "/") {
+		pathIdx = 1
+		leadSlash = true
+	}
+	// Remove the trailing slash segment if it exists
+	if strings.HasSuffix(req.URL.Path, "/") {
```
Comment (keithpaterson) [inline host.go]: does it make sense to have `ResourceNames(pathSegs)` take care of this instead of putting it on the caller?  I don't think we have a case where we want to keep the trailing "/" do we?
Also you might consider having it ignore the lead slash as well?

File: host_test.go [hunk 6]
```diff
@@ -37,6 +38,70 @@ var _ = Describe("Host", func() {
 			Expect(caHost.Logger()).To(BeNil())
 		})
 	})
+	Describe("Serving Basics", func() {
+		It("should receive all ids in instanced sub-resources", func() {
+			path := "/resource/id-123/other"
+			segments := ResourceNames(strings.Split(path, "/"))
+			// split sets the first elment to an empty string if the path starts with a slash
+			res := NewAPI(segments[1])
```
Comment (keithpaterson) [inline host_test.go:46]: I would prefer that this explicitly names the new API instance "resource" instead of relying on the output of `ResourceNames()`; it more clearly shows & tests the relationship between the resource and the path.

so `res := NewAPI("resource")` here.

Obviously this applies to other tests as well.

File: host_test.go [hunk 7]
```diff
@@ -37,6 +38,70 @@ var _ = Describe("Host", func() {
...
+			res.SetID(`id-[0-9]+`)
+			subRes := NewAPI(segments[3])
```
Comment (keithpaterson) [inline host_test.go:48]: I believe this would be `subRes := NewAPI("other")`, assuming my previous comment is accepted?

File: host_test.go [hunk 8]
```diff
@@ -37,6 +38,70 @@ var _ = Describe("Host", func() {
...
+				Expect(ctx.Value(Key(fmt.Sprintf("id_%s", res.Name())))).To(BeNil())
+				Expect(ctx.Value(Key(fmt.Sprintf("id_%s", resources.Name())))).To(Equal("id-123"))
+				Expect(ctx.Value(Key(fmt.Sprintf("id_%s", subRes.Name())))).To(BeNil())
+			})
+			req := httptest.NewRequest(http.MethodGet, path, nil)
+			caHost.Serve(httptest.NewRecorder(), req)
+			Expect(handlerCalled).To(BeTrue())
+
+		})
+		It("should store an empty path segments slice if no segments exist", func() {
```
Comment (keithpaterson) [inline host_test.go:119]: devil's advocate: I'd like tests with three-resource paths:
"/resource/id-123/camp/id-456/other"
"/resource/id-123/camp/id-456/other/id-789"

Would this also be supported?
"/resource/id-123/camp/other/id-789"

I'm not sure; it feels like a bad idea but maybe there's a not-horrible reason for something like this?

File: api_test.go [hunk 9]
```diff
@@ -372,13 +319,61 @@ var _ = Describe("Api", func() {
...
+			Expect(err).To(MatchError(resweave.ErrNilResource))
+		})
+		It("should error if the sub-resource is already added", func() {
+			res := resweave.NewAPI("foo")
+			subRes := resweave.NewAPI("bar")
+			err := res.AddSubResource(subRes)
+			Expect(err).ToNot(HaveOccurred())
+			err = res.AddSubResource(subRes)
+			Expect(err).To(MatchError(resweave.ErrResourceAlreadyExists))
+		})
```
Comment (keithpaterson) [inline api_test.go:351]: devil's advocate again: should we test multiple sub-resources as well as a longer parent-child chain e.g. "foo/1/bar/2/baz/3" ?

File: api_test.go [hunk 10]
```diff
@@ -372,13 +319,61 @@ var _ = Describe("Api", func() {
...
+			err := res.AddInstancedSubResource(nil)
+			Expect(err).To(MatchError(resweave.ErrNilResource))
+		})
+		It("should error if the sub-resource is already added", func() {
+			res := resweave.NewAPI("foo")
+			subRes := resweave.NewAPI("bar")
+			err := res.AddInstancedSubResource(subRes)
+			Expect(err).ToNot(HaveOccurred())
+			err = res.AddInstancedSubResource(subRes)
+			Expect(err).To(MatchError(resweave.ErrInstancedResourceAlreadyExists))
```
Comment (keithpaterson) [inline api_test.go]: also here: test multiple child resources and some longer resource chains

File: api_test.go [hunk 11]
```diff
@@ -372,13 +319,61 @@ var _ = Describe("Api", func() {
...
+		})
+	})
 })
 
-func resultsMatch(expStatusCode int, res resweave.Resource, req *http.Request) ([]byte, error) {
+func resultsMatch(expStatusCode int, inputContext context.Context, res resweave.Resource, req *http.Request) ([]byte, error) {
```
Comment (keithpaterson) [inline api_test.go]: hm, this isn't quite "resultsMatch()", more like "verifyResponseCode()"?
also is there value in adding a test that the body is what we expect?  if nil body is passed in we should be getting a nil/empty body in the response?  (and then we can name this "verifyResponse()"

but I haven't investigated all the calls you have for this.
.. but that would allow us to use this like a matcher and not need any returned values.  Either it fails via `Expect(...)` or it passes.

File: api.go [hunk 12]
```diff
@@ -78,30 +83,58 @@ func (i ID) Find(s string) (string, bool) {
...
-	actionMap actionFuncMap
-	id        ID
-	handler   HandlerFunction
+	name                  ResourceName
+	actionMap             actionFuncMap
+	id                    ID
+	handler               HandlerFunction
+	subResources          ResourceMap
+	instancedSubResources ResourceMap
 }
```
Comment (keithpaterson) [inline api.go:126]: This is in a funny spot but it won't let me comment on unchanged code so ...

Child resources may want/need access to the parent resource ID, so I think we need to add `GetResourceIDValue(ctx context.Context, ResourceName string) (string, error)` using the current implementation (mostly), and then we can implement
```
func (bar *BaseAPIRes) GetIDValue(ctx context.Context) (string, error) {
    return bar.GetResourceIDValue(ctx, bar.Name())
} 
```

**Review by keithpaterson — CHANGES_REQUESTED** [review]

File: api.go (hunk 2 above)
Comment (keithpaterson) [reply api.go]: Hmm, other ideas:

"foo/<id>/bar" => Parent(id)/Child
"foo/bar => Parent/Sibling  or maybe  Parent/[Friend|Cousin|Peer] ... something along those lines .. I'm still not quite keen on "Sub" anything.

I'm trying to imagine it in documentation/conversation, and the thought is roughly
"<this> resource has a Child relationship with <parent> resource" -> implies the "parent/<id>/child" pattern

vs.

"<this> resource has a <???> relationship with <parent> resource" -> implies the "parent/???" pattern.

I think giving distinct names will make it more clear overall.

File: api.go (hunk 3 above)
Comment (mortedecai) [reply api.go]: Moving to resources

File: api.go (hunk 1 above)
Comment (mortedecai) [reply api.go]: Modified to AddChildResource as agreed.

File: api.go (hunk 12 above)
Comment (mortedecai) [reply api.go:126]: I agree they will need access to it.~; I'll think on adding it now, as it's mostly a syntactic sugar piece as it is accessible via the Context.  I may suggest adding it as a quick follow just to get the base functionality in.~

Added.

File: host_test.go (hunk 6 above)
Comment (mortedecai) [reply host_test.go:46]: I disagree that it shows it more clearly.

This is directly taking a segment from the parsed path. 

The integration test for hello world follows the paradigm you are suggesting, I don't see what we gain by changing this.

File: host_test.go (hunk 8 above)
Comment (mortedecai) [reply host_test.go:119]: I thought I had done one; my bad.

Comment (mortedecai) [reply host_test.go:119]: Added above this comment.

File: api_test.go (hunk 9 above)
Comment (mortedecai) [reply api_test.go:351]: Sorry resolved by accident.

The subresource addition is not recursive; it would be re-testing the same method again just on a new resource.

File: api_test.go (hunk 11 above)
Comment (mortedecai) [reply api_test.go]: The test for the expected body is done in the actual test case.  I didn't do the passing in of the test data, because there are some cases where the result didn't matter, it was more that we got something and got the expected status code back.

Comment (mortedecai) [reply api_test.go]: Renamed to verifyStatusGetBody

File: api_test.go (hunk 10 above)
Comment (mortedecai) [reply api_test.go]: Already tested via the host call.

File: host.go (hunk 5 above)
Comment (mortedecai) [reply host.go]: I don't think so.  ResourceNames is a utility method to just get back a set of ResourceName instances from a slice of strings.

To do that conversion, it would need to know the intent of the caller.

File: api_test.go (hunk 9 above)
Comment (keithpaterson) [reply api_test.go:351]: maybe add one test to prove that it can handle the longer chain, but you don't need to test all the functionality, just one of the simpler ones?

File: host.go (hunk 5 above)
Comment (keithpaterson) [reply host.go]: agree; I think I misread what was going on.
but it does feel weird to convert the values then check to see if we want the last one... would it make more logical sense to move the slash-checks up and have them (possibly) modify pathSegs instead?  Then you are only passing the actual segments you want converted into ResourceNames.   I suspect that's easier to follow:
```
pathSegs := strings.Split(req.URL.Path, "/")
if strings.HasSuffix(req.URL.Path, "/") {
  pathSegs = pathSegs[:len(pathSegs)-1]
}
...
reqPaths := ResourceNames(pathSegs)
```

It puts the slice manipulation on the source data rather than the result?
I'm probably just being pedantic but it feels better to me to "fix" the source info before conversion than to do the conversion and throw something away after.

File: host_test.go (hunk 6 above)
Comment (keithpaterson) [reply host_test.go:46]: I'm saying that given the path "A/id/B", you should be setting up `res` and `subRes` with the explicit names that you are expecting from that path.  And if the handler has problems finding those resources then we can fix it.

basically you're duplicating what's going on under the covers.  The test should imply knowledge of what it things is going on but proves that by not trying to do the same logic ... I feel like I'm not explaining it right.

It's the difference between "given this path I'll construct a res and a subres from it and run a test" vs. "given this path, I expect to get a res("this") and a subres("that"), and if those are wrong then we have a bug".

If this was a table test and the path was passed in I would likely be ok with the parsing but since path is explicitly stated here I think we should also explicitly state what we expect host to be doing, rather than peeling apart the path in code.

**Review by keithpaterson — CHANGES_REQUESTED** [review]
I still have a couple of concerns.

File: host.go (hunk 5 above)
Comment (mortedecai) [reply host.go]: A good compromise is one where no one is happy. ;)
Cut out the middle vars, streamlined it a bit.

File: host_test.go (hunk 6 above)
Comment (mortedecai) [reply host_test.go:46]: I understand what you're saying.

I disagree that we gain anything by doing this, as the exact paradigm you suggest is used in the integration test, from end to end.

My argument is that the integration tests cover what you're suggesting; that it's not worth the time to change this here, at this moment, and that, arguably, keeping this covers multiple methods of testing this, so that should the underlying implementation change, we have this method still tested, as well as the one you are asking for tested in the integration tests.

File: api.go [hunk 13]
```diff
@@ -45,10 +45,10 @@ const (
...
 
 var (
-	ErrIDNotFound                     = errors.New("no ID found")
-	ErrNilResource                    = errors.New("cannot add a nil resource")
-	ErrResourceAlreadyExists          = errors.New("sub-resource already exists")
-	ErrInstancedResourceAlreadyExists = errors.New("instanced sub-resource already exists")
+	ErrIDNotFound                 = errors.New("no ID found")
+	ErrNilResource                = errors.New("cannot add a nil resource")
+	ErrResourceAlreadyExists      = errors.New("sub-resource already exists")
+	ErrChildResourceAlreadyExists = errors.New("child sub-resource already exists")
```
Comment (keithpaterson) [inline api.go]: nit: "child resource ..."

**Review by keithpaterson — DISMISSED** [review]
one nit, and you seem to have a couple of check-errors to look at, otherwise I'm fine with this as-is.

(notwithstanding our discussion about being more explicit about names of things in the unit tests but it's not so important that I feel it should hold things up... I'd rather discover issues via unit tests, not integration tests)
